### ISSUE TITLE
Few comments

### ISSUE THREAD

**Issue by gytis-ivaskevicius** [NorfairKing/nix-does-not-guarantee-reproducibility#1]
I find it difficult to accept most of your arguments, I think the point of this writeup should have been 'Nix is reproducible or fails if reproducibility is impossible'. Nix should not be held responsible for not being able to bypass China firewalls, or for not being able to buy you a new computer, or for not contacting your ISP just because you don't have an internet connection, etc

**Poor point:**
> Nix could mitigate this issue by running all builds on a single core, but that should not because that would slow down builds massively. It also wouldn't necessarily help because running a build on one core does not prevent GHC from spawning multiple green threads anyway.

This would not resolve anything since CPU scheduling is not deterministic

**Points you should have mentioned:**
- `echo date > $out` And other common cases (Kinda addressed it)
- Kernel - it's one of the few impurities of Nix builds. Different kernel builds may handle syscalls differently
- File system - Another impurity that Nix does not handle
- Hardware. Example: `gcc -march=native`

**Comment by NorfairKing** [comment]
PR welcome

**Comment by gytis-ivaskevicius** [comment]
Can I remove some of your points in the PR? :D

**Comment by NorfairKing** [comment]
> Can I remove some of your points in the PR? :D

Depends if you want it to be merged, lol.

Keep in mind that "It's impossible to guarantee X" is even MORE of a reason not to lie about nix guaranteeing X, not less of one.

**Comment by gytis-ivaskevicius** [comment]
Yes, I agree but I don't think that presentation was done right. It sounds like you expect Nix to buy you a new computer

**Comment by arvindpunk** [comment]
Certain points seem rather absurd to me, nix itself is not failing in the examples.

Derivations have dependencies on hosted third party software (in this case, sources/binary files). Like @gytis-ivaskevicius mentioned, it's not nix at fault here that the 3rd party hosted software is unavailable/unreachable. You could theoretically write the entire 3rd party software/binary within the nix derivation itself, but that would basically mean hosting it in nixpkgs repo. Basically, this would be an impossible problem to solve, using nix or not.

Points raised by Gytis make more sense to me, the dependency on hardware arch, FHS, the kernel etc. all of which are inherently impure in nature.

Also, having randomness during build seems like an incorrect way to approach a problem (secret generation during builds?). Wouldn't it make more sense to do that during run time? (I could be wrong in this opinion, I'd be more than happy to be educated more on this).

---
update: I'm not the greatest nix user, still learning and trying to get the hang of it. If I claimed something wrong - I'd gladly be part of the discussion to learn more. cheers!

**Comment by NorfairKing** [comment]
@gytis-ivaskevicius and @arvindpunk I'm happy to add elaborations on other ways that you might make a build un-reproducible according to the definition listed in the corresponding section. PR welcome.


> It sounds like you expect Nix to buy you a new computer

I'm not so happy about having my position misrepresented.
All I'm suggesting is: "Let's please not lie to people."
